# **Inference**

In [16]:
import pickle
from tensorflow.keras.models import load_model
from tensorflow.keras.preprocessing.sequence import pad_sequences
import numpy as np

# Load tokenizer
with open('tokenizer.pkl', 'rb') as handle:
    tokenizer = pickle.load(handle)

# Load models
model_lstm = load_model("model_lstm.h5")
model_gru = load_model("model_gru.h5")
model_cnn = load_model("model_cnn.h5")


In [17]:
# Label mapping (ubah sesuai hasil LabelEncoder jika berbeda)
label_map = {
    0: "Negatif",
    1: "Netral",
    2: "Positif"
}

# Tiga ulasan untuk prediksi
ulasan = [
    "Aplikasi ini sangat membantu dan tampilannya menarik sekali!",
    "Cukup baik, tapi ada kendala saat masuk aplikasi.",
    "Saya kecewa, aplikasinya sering error dan lambat.",
    "Aplikasinya bagus, nyaman dan juga, sangat mudah digunakan dan transfer cukup cepat",
    "Keren , Saya sudah lama menggunakan aplikasi Dana, tidak ada kendala"
]


In [18]:
# Ubah teks ke token dan lakukan padding
sequences = tokenizer.texts_to_sequences(ulasan)
padded = pad_sequences(sequences, maxlen=100, padding='post', truncating='post')


In [19]:
# Fungsi untuk prediksi
def prediksi_sentimen(model, data, model_name):
    hasil = model.predict(data)
    label_pred = np.argmax(hasil, axis=1)
    print(f"\n=== Prediksi dengan {model_name} ===")
    for i, label in enumerate(label_pred):
        print(f"Ulasan: {ulasan[i]}")
        print(f"Prediksi Sentimen: {label_map[label]}")
        print(f"Probabilitas: {hasil[i]}")
        print("-" * 50)


In [20]:
# Prediksi dengan ketiga model
prediksi_sentimen(model_lstm, padded, "LSTM")
prediksi_sentimen(model_gru, padded, "GRU")
prediksi_sentimen(model_cnn, padded, "CNN")

1/1 ━━━━━━━━━━━━━━━━━━━━ 1s 568ms/step

=== Prediksi dengan LSTM ===
Ulasan: Aplikasi ini sangat membantu dan tampilannya menarik sekali!
Prediksi Sentimen: Netral
Probabilitas: [0.0163783  0.96309614 0.02052552]
--------------------------------------------------
Ulasan: Cukup baik, tapi ada kendala saat masuk aplikasi.
Prediksi Sentimen: Netral
Probabilitas: [0.08459176 0.77721953 0.13818873]
--------------------------------------------------
Ulasan: Saya kecewa, aplikasinya sering error dan lambat.
Prediksi Sentimen: Negatif
Probabilitas: [9.9914086e-01 2.6967116e-05 8.3220529e-04]
--------------------------------------------------
Ulasan: Aplikasinya bagus, nyaman dan juga, sangat mudah digunakan dan transfer cukup cepat
Prediksi Sentimen: Positif
Probabilitas: [3.851822e-02 6.940727e-04 9.607877e-01]
--------------------------------------------------
Ulasan: Keren , Saya sudah lama menggunakan aplikasi Dana, tidak ada kendala
Prediksi Sentimen: Netral
Probabilitas: [0.1950503  0.52

# 📊 Analisis Prediksi Sentimen Aplikasi DANA

## ✅ Ringkasan Hasil Prediksi

| No | Ulasan                                                                                   | Prediksi LSTM | Prediksi GRU | Prediksi CNN |
|----|-------------------------------------------------------------------------------------------|---------------|---------------|---------------|
| 1  | Aplikasi ini sangat membantu dan tampilannya menarik sekali!                             | Netral        | Netral        | Netral        |
| 2  | Cukup baik, tapi ada kendala saat masuk aplikasi.                                        | Netral        | Netral        | Netral        |
| 3  | Saya kecewa, aplikasinya sering error dan lambat.                                         | Negatif       | Negatif       | Negatif       |
| 4  | Aplikasinya bagus, nyaman dan juga sangat mudah digunakan dan transfer cukup cepat       | Positif       | Positif       | Positif       |
| 5  | Keren, saya sudah lama menggunakan aplikasi Dana, tidak ada kendala                      | Netral        | Positif       | Netral        |

---

## 📌 Penjelasan Per Ulasan

### 🔹 1. **"Aplikasi ini sangat membantu dan tampilannya menarik sekali!"**
- **Prediksi:**
  - LSTM: `Netral` (Prob: `0.9630`)
  - GRU: `Netral` (Prob: `0.8969`)
  - CNN: `Netral` (Prob: `0.8799`)
- **Catatan:** Meskipun bernada positif, tidak cukup kuat untuk digolongkan sebagai "Positif" oleh model.

---

### 🔹 2. **"Cukup baik, tapi ada kendala saat masuk aplikasi."**
- **Prediksi:**
  - LSTM: `Netral` (Prob: `0.7772`)
  - GRU: `Netral` (Prob: `0.7985`)
  - CNN: `Netral` (Prob: `0.9437`)
- **Catatan:** Kalimat campuran (positif dan negatif) membuat ketiga model sepakat memilih sentimen Netral.

---

### 🔹 3. **"Saya kecewa, aplikasinya sering error dan lambat."**
- **Prediksi:**
  - LSTM: `Negatif` (Prob: `0.9991`)
  - GRU: `Negatif` (Prob: `0.9999`)
  - CNN: `Negatif` (Prob: `1.0000`)
- **Catatan:** Semua model sangat yakin bahwa ini adalah sentimen Negatif karena ekspresi kekecewaan yang jelas.

---

### 🔹 4. **"Aplikasinya bagus, nyaman dan juga sangat mudah digunakan dan transfer cukup cepat."**
- **Prediksi:**
  - LSTM: `Positif` (Prob: `0.9607`)
  - GRU: `Positif` (Prob: `0.9975`)
  - CNN: `Positif` (Prob: `0.9921`)
- **Catatan:** Sentimen positif sangat kuat karena kata-kata pujian dan pengalaman baik.

---

### 🔹 5. **"Keren, saya sudah lama menggunakan aplikasi Dana, tidak ada kendala."**
- **Prediksi:**
  - LSTM: `Netral` (Prob: `0.5217`)
  - GRU: `Positif` (Prob: `0.4286`)
  - CNN: `Netral` (Prob: `0.6407`)
- **Catatan:** Kalimat positif ringan. GRU lebih responsif terhadap ekspresi positif ringan dibanding LSTM dan CNN.

---

## 🧠 Kesimpulan

- **LSTM** dan **CNN** cenderung konservatif dan lebih memilih Netral untuk ulasan dengan nada ambigu.
- **GRU** sedikit lebih sensitif terhadap ekspresi halus positif, bisa mengklasifikasikan Positif meskipun tanpa kata pujian eksplisit.
- Semua model menunjukkan **konsistensi tinggi** terhadap ulasan yang jelas-jelas Positif atau Negatif.

---

